---
title: "SLIDE"
output: rmarkdown::pdf_document
vignette: >
  %\VignetteIndexEntry{SLIDE}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include = FALSE}
library(knitr)

knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
   knitr.graphics.auto_pdf = TRUE, nobreak = TRUE)

knitr::opts_knit$set(root.dir = '..')
```

**SLIDE Pipeline**

The SLIDE pipeline is consisted of below two steps. We recommend to run both steps on a computational cluster for optimal computational time.

1.  Calculating and select latent factors (LFs) for multiple input parameter combinations.

2.  Reviewing the output of step 1 and choose the optimal parameters for rigorous k-fold CV.

```{r workflow, eval = F, fig.cap = "SLIDE workflow", out.width="60%", fig.align="center"}
knitr::include_graphics('build_files/Vignette.png')
```
# **Step 1: Parameter Tuning**

## **1-1 Pre-Processing and Checking Input Data**

Due to the assumption-free-nature of SLIDE, there are no data modality limit to the input of SLIDE.

[The key input to SLIDE are just two csv files, the data matrix x and the response vector y file post pre-processing such as batch effect correction and/or normalization (for example, for scRN-seq data, first process with the standard Seurat pipeline).]{.underline}

The x file contains your data in a sample by feature format, such as single cell transcriptomics (**cell by gene**), or spatial proteomics (**region by protein**). The y file contains the responses of the data, such as severity of disease, spatial regions or clonal expansion. Since SLIDE is a regression method, when the response vector has multiple unique values (not just two classes), please make sure there are an ordinal relation between the y values. **Please make sure both csv files have row names and column names.**

In this tutorial, we are going to use the example Systemic Sclerosis dataset we have used in the SLIDE paper. This dataset is a human skin cell scRNA-seq dataset that we have transformed into the pseudo-bulk format.

[If you have human single-cell data,]{.underline} our recommended workflow is to pseudobulk your dataset since the cell-to-cell variability is high. [If you have mouse single cell data]{.underline}, with the reduced cell-to-cell variability, you can consider each cell as an sample. For single-cell datasets, the sparsity might be extremely high with high feature numbers. In this case, please see below for how to reduce the number of features and samples with large amount of zeros.  
 
 
 
 
## **1-2 Parameters**

The input to SLIDE is the path to a YAML file which documents the parameters.

SLIDE accepts many parameters from user input to give user as much freedom to tweak the method to their own data as possible. We have set default values to many parameters that works well with majority of the data moralities. Here, we explain what each of these parameters mean.

```{r load_example_data_class, eval = F, fig.cap = "Example Yaml (Continuous Y)", out.width="60%", fig.align="center"}
knitr::include_graphics('vignettes/build_files/test_yaml_example.png')
```
```{r load_example_data_binary, eval = F, fig.cap = "Example Yaml (Binary Y)", out.width="60%", fig.align="center"}
knitr::include_graphics('vignettes/build_files/yaml_example.png')
```
**x_path:** a string of the path to the data matrix (x) in the csv format [with row names and column names where each row is samples (cells, patients, regions...) and each column is a features (genes, proteins...)]{.underline}.

**y_path:** a string of the path of the response vector (y) in the csv format [with row names and column names where each row is a sample and the column would be the outcome of interest.]{.underline}.

**out_path:** a string of the path of a folder to store all output files (please see below section to interpret the outputs).

**delta:** control the number of all latent factors. The higher the delta, the less number of latent factors will be found. Default as 0.01 and 0.1.

**lambda:** control the sparsity of all latent factors. The higher the lambda, the less number of features will be in a latent factor. Default as 0.5 and 1.

**spec:** control the number of significant latent factors. The higher the spec, the less number of significant latent factors will be outputted. [The desired number of output should be between 5 to 12 LFs]{.underline}. Default as 0.1.

**y_factor:** set to false if not binary and true if binary.

**y_levels:** null if y is continuous or ordinal. If y is binary, input a list of the correct order relationship such as [0, 1] or [1, 2].

**eval_type:** the performance evaluation metric used. corr for continuous Y and auc for binary Y.

**SLIDE_iter:** the number of times to repeat the SLIDE latent factor selection algorithm. The higher the iteration, the more stable the performance would be. Default as

**SLIDE_top_feats:** the number of top features to plot from each latent factor. If set as n, a union of the top n weighted features and top n correlated (with y) features will be outputted.

**do_interacts (optional):** set to false if don't want interacting latent factors. Default as TRUE

**thresh_fdr:** set to lower if co-linearity of the features in the data matrix is high. Default as 0.2.
  
  
  
## **2-1 Step1 of SLIDE Framework**

Once your YAML file is ready, we first recommend using a YAML validator website to ensure your YAML file is correctly formatted.

We can then check if the YAML file is read in correctly by reading it in as an variable.

```{r initial, message = F}
library(SLIDE)
yaml_path = "examples/test.yaml"
input_params = yaml::read_yaml(yaml_path)

knitr::kable(data.frame(arg = unlist(input_params)))
```
**If you have a sparse dataset such as scRNA-seq datasets**, we recommend filtering out samples and features that have too many zeros. zeroFiltering function will remove samples with more than g_thresh number of zeros and features with more than c_thresh number of zeros. An appropriate data matrix should at most have around 3-4k features. See `Preprocessing-and-Filtering` vignette for more information.

<!-- # ```{name1} -->
<!-- # #zeroFiltering(yaml_path, g_thresh, c_thresh) -->
<!-- # ``` -->

We then check if your data files are formatted correctly.

```{r name2, warning=FALSE, message=FALSE}
checkDataParams(yaml_path)
```

If everything is formatted correctly, you can now run [Step1]{.underline} of SLIDE. If sink_file set to TRUE, all print statement will be printed to a txt file.

```{r name3, warning=FALSE}
 optimizeSLIDE(input_params, sink_file = FALSE)
```

<!-- ## **2-2 Interpret Step1 Output Files** -->

<!-- The output files are now populated to the output folder we specified in the YAML file. -->

<!-- First investigate the summary_table.csv.  -->
